TECH-4409 port build_hash and star operator from hobosupport

mikeweaver opened the pull request.

mikeweaver committed:
TECH-4409 port build_hash and star operator from hobosupport

mikeweaver commented on lib/invoca/utils/array.rb:
```diff
@@ -0,0 +1,17 @@
+# Invoca ::Array extensions
+class Array
+
+  alias_method :multiply, :*
```
Might need to pull in https://github.com/Invoca/hobosupport/blob/master/lib/hobo_support/fixes/module.rb#L7 to make this safe. I will add it if needed after testing these modifications with other gems.

ColinDKelley replied:
Oh, I hope not! I think we can walk away from `alias_method_chain` and its variations. They shouldn't be necessary now that we have `prepend`...which is _so_ much cleaner.
Can you recast this to use `prepend`? Something like:
```
module Invoca
  module Utils
    module ArrayMultiply
      def *(rhs = nil)
        if rhs
          super
        else
          MultiSender.new(self, :map)
        end
      end
    end
  end
end

Array.prepend(Invoca::Utils::ArrayMultiply)
```
(Foreshadowing a comment below... that you can move `MultiSender` out of the `Enumerable` namespace and into ours.)

ColinDKelley commented on CHANGELOG.md:
```diff
@@ -0,0 +1,12 @@
+# CHANGELOG for `invoca-utils`
```
💯 Thanks for adding this!

mikeweaver replied:
It is a nice change of pace to work on code. :-)

ColinDKelley commented on CHANGELOG.md:
```diff
@@ -0,0 +1,12 @@
+
+Note: This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [0.2.0] - 2020-04-24
```
BTW our convention has evolved to putting `Unreleased` here during the review, and then committing an update with the date just before we run `gem release`.
Not that you need to change it here now... just remember to bump this date and commit when you run the `gem release`.

mikeweaver replied:
Good to know. Will do!

ColinDKelley commented on CHANGELOG.md:
```diff
@@ -0,0 +1,12 @@
+- Enumerable::build_hash method ported from HoboSupport
+- Enumerable::* operator ported from HoboSupport
+
+[0.2.0]: https://github.com/Invoca/process_settings/compare/v0.2.0...v0.1.1
```
I believe those should be reversed:
```
[0.2.0]: https://github.com/Invoca/process_settings/compare/v0.1.1...v0.2.0
```

ColinDKelley commented on lib/invoca/utils/enumerable.rb:
```diff
@@ -0,0 +1,31 @@
+
+  class MultiSender
+
+    undef_method(*(instance_methods.map { |m| m.to_s } - %w*__id__ __send__ object_id*))
```
Gah, that's crazy they used `%w*`! What were they thinking? `%w[ ... ]` makes it so much more clear that it's a (space-separated) array:
```
%w[__id__ __send__ object_id]
```
But even better: starting with Ruby 1.9, the instance methods are symbols, we can avoid the strings and `%w` entirely:
```
undef_method(*(instance_methods - [:__id__, :__send__, :object_id]))
```

mikeweaver replied:
Yeah, I didn't even want to touch that line when I saw it. Wasn't worth research time to fix it. But, hey, if you know how to do it off the top of your head, why not!

ColinDKelley commented on lib/invoca/utils/enumerable.rb:
```diff
@@ -0,0 +1,31 @@
+    res
+  end
+
+  class MultiSender
```
I think this should also be a module in `Invoca::Utils`. No need to clutter `module Enumerable` with it.

mikeweaver replied:
Moved. Would you test this independently? Or rely upon the Array tests?

ColinDKelley replied:
I like 1 test using Array, and 1 of our own class that includes Enumerable. That covers the 2 extremes.

mikeweaver replied:
Done.

ColinDKelley commented on lib/invoca/utils/enumerable.rb:
```diff
@@ -0,0 +1,31 @@
+
+  end
+
+  def *()
```
Huh, any idea when this gets called? It's a special case vs. `method_missing` above being general. And it doesn't allow any arguments.

Update: I think this is how they're enabling chaining. If so, can you add a test of this? As in:
```
assert_equal ["OM", "HO", "OR"], ['some', 'short', 'words'].*.upcase.*.slice(1, 3)
```

mikeweaver replied:
This is how the .* functionality gets added to Enumerable classes other than Array. These did a special patch for Array, I suppose, to preserve the multiplication behavior. Leaving this here could break other Enumerables multiplication behavior...unless I implement it like they did in Array. Which, now that I am typing this, seems like a good idea.

mikeweaver replied:
Actually, no, that won't work of course! We are not pre-pending *() onto any of the classes, so the call to super makes no sense. Maybe we should just dump the support for Enumerables in general. I would imagine we only use for Arrays. And, we will get a runtime error (likely to be found in tests) if we are using it on other Enumerables. I'm going to kill it for now.

ColinDKelley replied:
Oh, weird.

I agree that it doesn't seem to make sense to preserve the existing `*` behavior only for `Array`--that seems like it would work fine, generally.

Also, I doubt we need that `:multiply` alias. I'm 99.9% sure no one knew it was there.

mikeweaver replied:
I was actually making the opposite point. I only want to add the new * behavior to Array and leave all other enumerables untouched.

ColinDKelley commented on test/unit/array_test.rb:
```diff
@@ -0,0 +1,16 @@
+  context '* operator' do
+    should 'call the same method on each item in an array and return the results as an array' do
+      assert_equal([4, 5, 5], ['some', 'short', 'words'].*.length)
+    end
```
I think you should also test a method with some arguments, like
```
assert_equal(["om", "ho", "or"], ['some', 'short', 'words'].*.slice(1, 2)
```

mikeweaver replied:
Good idea.

ColinDKelley commented on test/unit/enumerable_test.rb:
```diff
@@ -0,0 +1,38 @@
+class EnumerableTest < Minitest::Test
+
+  context 'build_hash' do
+    should 'convert arrays of two items to hash key/value pairs' do
```
arrays of [key, value] to hash of key => value

mikeweaver replied:
That reads better!

ColinDKelley commented on test/unit/enumerable_test.rb:
```diff
@@ -0,0 +1,38 @@
+    end
+
+    should 'ignore nils returned from map' do
+      assert_equal({ 'some' => 4, 'words' => 5 }, ['some', 'short', 'words'].build_hash { |s|  s == 'short' ? nil : [s, s.length] })
```
The above actually tested `nil` also. Probably just remove the `unless` up there so it's the simplest, most common case?

mikeweaver replied:
Good point.

ColinDKelley commented on test/unit/enumerable_test.rb:
```diff
@@ -0,0 +1,38 @@
+    should 'convert empty arrays into nil keys and values' do
+      assert_equal({ 'some' => 4, nil => nil, 'words' => 5 }, ['some', 'short', 'words'].build_hash { |s| s == 'short' ? [] : [s, s.length] })
+    end
+  end
```
These are bizarrre! I'm tempted to not port across. I seriously doubt anyone is using that weirdness.
Or a middle ground, we could use ActiveSupport::Deprecate on those. But it's probably not necessary.

ColinDKelley replied:
I suppose if we don't port across, those weird cases would all raise `ArgumentError`. The only cases we need to allow--IMO--are
1. a `[key, value]` pair
2. `nil`

mikeweaver replied:
Who knows if we are relying on this oddball behavior. If we are, it probably isn't covered by tests, so we are going to find out the hard way. I'd rather leave it alone.

ColinDKelley replied:
Ok!

ColinDKelley commented on test/unit/enumerable_test.rb:
```diff
@@ -0,0 +1,38 @@
+
+  context '* operator' do
+    should 'call the same method on each item in an enumeration and return the results as an array' do
+      assert_equal([4, 5, 5], Set['some', 'short', 'words'].*.length)
```
Odd that this uses `Set`. How about making this our own class (just in this test) that mixes in `Enumerable`, to prove that it works with `Enumerable` in general?

mikeweaver replied:
I was simply trying to test that the *() function works on other enumerable types. Is making our own just for test purposes really worth it? I could instead, assert that Set inherits from Enumerable. What do you think of that?

mikeweaver replied:
I removed support for enumerables and only kept Array

ColinDKelley reviewed (changes requested):
Thanks for doing this.

Man, hobo has some seriously-too-clever code!

mikeweaver committed:
Fix errors in CHANGELOG

mikeweaver committed:
Improve test names for enumerable module

mikeweaver committed:
Add additional test coverage for array and enumerable

mikeweaver committed:
Remove support for * function from enumerable

Without using prepend, overriding * is actually dangerous as it changes the multiplication behavior.

mikeweaver committed:
Move MultiSender class to its own file, use prepend instead of alias_method

mikeweaver commented:
@ColinDKelley Ready for re-review

mikeweaver committed:
Add tests for MultiSender class

ColinDKelley reviewed (commented):
Just one comment left:

> I was actually making the opposite point. I only want to add the new * behavior to Array and leave all other enumerables untouched.

But it formerly did work with any `Enumerable`. I think we are probably using that, and want to keep it. For example, this currently works fine on `web` `master`
```
irb> {region: 'east', protocol: 'https'}.*.join(':')
["region:east", "protocol:https"]
```
So I think we want the module to be mixed into Enumerable too. Definitely add a test for the above!
Hmm, this is disturbing:
```
module Boom
  def boom
    raise 'boom!'
  end
end

class Before
  include Enumerable
end

module Enumerable
  prepend Boom
end

class After
  include Enumerable
end

After.new.boom
=> RuntimeError: boom!

Before.new.boom
=> NoMethodError: undefined method `boom' for []:Array
```
I didn't realize that nested include/prepend like that isn't retroactive. Any class that already had included `Enumerable` won't get the new mix-in--only new ones will.

By comparison, adding methods to `Enumerable` itself _is_ retroactive:
```
module Enumerable
  def boom
    raise '!!!'
  end
end
```
Seems like we're stuck with mutating `Enumerable`. :(

mikeweaver committed:
Re-add support for * operator to Enumerable

It turns out we DO use this for Enumerables other than Array

mikeweaver committed:
Address code style nits

The Style/MethodMissingSuper and Style/MissingRespondToMissing don’t apply to this usage of method_missing. We want the method missing exception to come from the wrapper enumerable object, not the MethodMissing class. Thus, callign Super makes no sense.

We could implement respond_to, but, I can’t imagine someone wanting to call [1,2,3].*.responds_to(…)

mikeweaver committed:
Prep for pre-release gem publishing

ColinDKelley reviewed (approved):
Looks great!

mikeweaver committed:
Release version 0.2.0
